### ISSUE TITLE
Repairnator for Jenkins: run repairnator when PR are coming or builds fail

### ISSUE THREAD

**Issue by surli** [eclipse/repairnator#469]
Configuration space:

* Trigger
  * BS: Build step for Repairnator
  * Post-build action of "Freestyle job"
    * PBA-GH For Github builds (incl. PR builder)
    * PBA-GIT For any Git repo
  * NJT: New full-fledged new job type "Repairnator job"   
* Feedback:
  * Feedback-Jenkins: Patch stored and shown on Jenkins directly
  * Feedback-email: Patch sent over email
  * Feedback-Github: Comment/Pull-request on Github
  * Feedback-Gitlab:  Comment/Pull-request on Gitlab

**Comment by surli** [comment]
I'm closing this one: I won't have time to fix the issues I had with jenkins. And we manage to work using only Travis.

**Comment by monperrus** [comment]
Many industry partners use Jenkins, it would be great to have a Jenkins plugin for Repairnator.

**Comment by monperrus** [comment]
The Jenkins plugin adds a new kind of "Post-build action", called "Run Repairnator upon failing builds".

**Comment by henry-lp** [comment]
@monperrus Hi Martin , I just recently start working on this per discussion . 

The plan:
* Create a dummy plugin with post build [done]
* Introduce `repairnator-pipeline` as maven dependency into the dummy plugin [WIP]
* Host the plugin on Jenkins central repository

This way if the plan is possible, we also ease the CI/CD for the plugin when newer version of pipeline itself is deployed, since it's only  a maven dependency specified in the pom.xml for the plugin. That's my current design .

**Comment by henry-lp** [comment]
Now to my current problem .

This issue happens when I tried to introduce `repairnator-pipeline` [maven dependency[(https://search.maven.org/artifact/fr.inria.repairnator/repairnator-pipeline/3.1/jar) into one of my dummy travis project [Henrylp/travisplay](https://github.com/henry-lp/travisplay) from Sonatype Nexus . Pom file and error output log are down below .

It seems that it unable to install the `repairnator-root` dependency inside the `repairnator-pipeline` dependency according to the err log . 

It would be nice If I know a fix or a workaround for this . Otherwise I will modify my plan a bit to avoid this :) .

My pom.xml
```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>apps</groupId>
  <artifactId>travisplay</artifactId>
  <packaging>jar</packaging>
  <version>1.8-SNAPSHOT</version>
  <name>travisplay</name>
  <url>http://maven.apache.org</url>
  <properties>
    <!-- https://maven.apache.org/general.html#encoding-warning -->
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    
        <maven.compiler.source>1.8</maven.compiler.source>
        <maven.compiler.target>1.8</maven.compiler.target>
    </properties>
  <repositories>
    <repository>
      <id>repairnator</id>
      <url>https://repo1.maven.org/</url>
    </repository>
  </repositories>
  <dependencies>
    <dependency>
      <groupId>junit</groupId>
      <artifactId>junit</artifactId>
      <version>4.11</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>commons-codec</groupId>
      <artifactId>commons-codec</artifactId>
      <version>1.11</version>
    </dependency>
    <dependency>
      <groupId>fr.inria.repairnator</groupId>
      <artifactId>repairnator-root</artifactId>
      <version>3.1</version>
      <type>pom</type>
    </dependency>
    <dependency>
      <groupId>fr.inria.repairnator</groupId>
      <artifactId>repairnator-pipeline</artifactId>
      <version>3.1</version>
      <classifier>jar-with-dependencies</classifier>
    </dependency>
  </dependencies>
  <build>
  </build>
</project>
```

Error output:
```
Downloading from repairnator: https://repo1.maven.org/fr/inria/repairnator/repairnator-root/$%7Brevision%7D/repairnator-root-$%7Brevision%7D.pom
Downloading from tdurieux-maven-repository-snapshot: https://tdurieux.github.io/maven-repository/snapshots/fr/inria/repairnator/repairnator-root/$%7Brevision%7D/repairnator-root-$%7Brevision%7D.pom
Downloading from tdurieux-maven-repository-release: https://tdurieux.github.io/maven-repository/releases/fr/inria/repairnator/repairnator-root/$%7Brevision%7D/repairnator-root-$%7Brevision%7D.pom
Downloading from tdurieux-maven-repository-release-github: https://github.com/tdurieux/maven-repository/tree/master/releases/fr/inria/repairnator/repairnator-root/$%7Brevision%7D/repairnator-root-$%7Brevision%7D.pom
Downloading from tdurieux-maven-repository-snapshot-github: https://github.com/tdurieux/maven-repository/tree/master/snapshots/fr/inria/repairnator/repairnator-root/$%7Brevision%7D/repairnator-root-$%7Brevision%7D.pom
Downloading from central: https://repo.maven.apache.org/maven2/fr/inria/repairnator/repairnator-root/$%7Brevision%7D/repairnator-root-$%7Brevision%7D.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.008 s
[INFO] Finished at: 2019-12-16T17:46:51+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project travisplay: Could not resolve dependencies for project apps:travisplay:jar:1.8-SNAPSHOT: Failed to collect dependencies at fr.inria.repairnator:repairnator-pipeline:jar:jar-with-dependencies:3.1: Failed to read artifact descriptor for fr.inria.repairnator:repairnator-pipeline:jar:jar-with-dependencies:3.1: Could not find artifact fr.inria.repairnator:repairnator-root:pom:${revision} in repairnator (https://repo1.maven.org/) -> [Help 1]
```

**Comment by henry-lp** [comment]
Update:
issue is fixed (I think , will report back if there are any other complications) . 
The fix: set repository url in pom.xml as `https://search.maven.org/`

**Comment by henry-lp** [comment]
@monperrus ,
Current issue:
Description: Unable to catch output from repairnator-pipeline and display it on Jenkins plugin Web UI . Jenkins has its own logger and I am unable to find a way to redirect our pipeline loggers to Jenkins logger either directly or indirectly first through standard output stdout.I is however possible to catch stdout lines printed out by System.out.println or err from repairnator to display it on Jenkins UI. 

Suggestions:
* Also system.out.println if it's for Jenkins plugin
* Change the current way of logging of all components to forward from System.out to logger instead . Then we can redirect System.out to Jenkins logger if we use it for the plugin.

**Comment by monperrus** [comment]
I'm pretty sure that's doable as is. The plugin can provide a specific "logback.xml" (different from
the default one in the repo) to the pipeline to output to system.out. See the documentation of
logback. WDYT?

**Comment by monperrus** [comment]
Very good plan.

We need two other steps:

- push the code on the repairnator repo
- set up CD: each time a commit touches it, it is redeployed on the Jenkins central repository

**Comment by henry-lp** [comment]
@monperrus Hi Martin :), Yes I checked around a bit and understood the theory behind it. I agree that It would succeed in the usual case, but yesterday after a quick look at Jenkins plugin mandatory parent dependency, I found out that It is not allowed to exclude their logging dependencies, so I went and find out that they use `java.util.logging` as their main logging to attempt to maybe override their configuration in parent in my plugin. I need to investigate more on this option. I also found a way to directly access their log file which log everything happened to Jenkins so we maybe have an option to pipe their log file with a buffer reader and put that on the UI.

Currently I have been busy with 2 other issues with the repairnator pipeline itself . Firstly, since we execute the pipeline as a dependency which is different than the usual context , we can't use certain classes like ServiceLoader to load in the tool and I think I solved this by making a method to manually load all the AbstractRepairSteps instead (The output is identical between the usual usecase and Jenkins, when running all the test cases, so I assume that it should be fine).

Secondly, It has been an issue with Git cloning step when running in the real Jenkins (not just `mvn hpi:run`), which I later found out that we can't create files without admin access as Jenkins and we can't just assume that we know the general file system structure of the machine we are running at. I just tested a solution using `com.google.common.io.Files.createTempDir()` and This work . So right now I am trying to integrate this somehow by refactoring the pipeline. 

Hopefully there are no serious bug left .

**Comment by henry-lp** [comment]
> Very good plan. We need two other steps: - push the code on the repairnator repo - set up CD: each time a commit touches it, it is redeployed on the Jenkins central repository

Ofcourse :) , I'll create a WIP later once I think I solved all the found functionalities issues on the real Jenkins platform.

**Comment by henry-lp** [comment]
@monperrus , Logging issue is now fixed . I managed to find and pipe all their logger to stderr and then I set stderr to the UI logger :) . Next is to fine-tuning this so that I only collect log from those loggers which output by our pipeline.

**Comment by monperrus** [comment]
> Logging issue is now fixed 

Congrats!

**Comment by henry-lp** [comment]
@monperrus hi Martin :) ,
Latest achievements:
Verified and validated by me at my Jenkins server:
* NPE fix using plugin.
* Auto detect branch from Jenkins native Git Plugin and other integration issues with this are solved.
* PR when patches are found [Link](https://github.com/surli/failingProject/pull/24) . 

Future features:
* Choice of other repair tools other than just NPEfix

Challenges:
* Mostly it's verfication and validation work which is tiresome. We are using repairnator in a new context so all the current features inside repairnator need to be re-verified and re-validated . Also new tests have to be written to ensure qualities

Todo:
* Plugin test for NPE fix
* Hosting plugin on Jenkins namespace.

**Comment by monperrus** [comment]
Nice progress! Good job Henry!

**Comment by henry-lp** [comment]
PR is ready to be merged as first version of the plugin.

**Comment by monperrus** [comment]
For the record, if we want to deploy on jenkins.io, we need to host the code of the plugin as a repo on the jenkinsci organization, see https://jenkins.io/doc/developer/publishing/requesting-hosting/

As alternative for now, we'll simply deploy HPI files to be used by our partners.

**Comment by henry-lp** [comment]
todo:
* Trigger:
  * BS: Build step for Repairnator
* Auto release .hpi file on Tduriex github repo

**Comment by henry-lp** [comment]
@monperrus hi Martin :) , there is an issue regarding making the plugin as sub module of repairnator, because it already has another parent `org.jenkins-ci.plugins` , there seems to be no work around due to the fact that we need to package the `hpi` instead of `jar` which specificly require the mentioned parent (so dependency injection with maven `dependency management` will not work).  My suggestion is to have this on the root level `same as maven repair` folder and deploy this separately.

**Comment by monperrus** [comment]
I think you can simply add as as a submodule with `<module>` i the root pom without adding the super
folder pom as parent pom.

**Comment by henry-lp** [comment]
@monperrus I see. I just make a PR . If no problem we can try merging that and see if there are any hpi on Sonatype :) .

**Comment by henry-lp** [comment]
@monperrus Deployment failed due to unable to submit .hpi to sonatype staging server . I guess we have to try a manual or auto release to Github instead. :) . This need to be reverted  or I can make a new PR to remove the new module in the pom.

**Comment by henry-lp** [comment]
I found out that we can deploy to Jenkins.ci repo instead , [link to hpi](https://repo.jenkins-ci.org/snapshots/fr/inria/repairnator/repairnator-plugins/) . we can try to merge the new fix PR .

**Comment by henry-lp** [comment]
Summary of the work:

Currently: repairnator-jenkins-plugin is only able to run NPEFix other tools fails due to the reasons below in Jenkins Context. Jenkins context generates, for instance no `.class` files like the usual execution context, it only use the classes directly as it is and files specified explicitly in the code , which also has a code to specificly tell how to use the file. Other than this is the dependency clash between required parent (must have for any jenkins plugin) which overwrites or invalidate the logging system used by repairnator or its transitive dependency. This is why some class does not work or imports are missed out completely during execution, since we did not actually compile repairnator directly as we did in the usual.

* AssertFixer - problem with `Spoon`
 * spoon.SpoonException: Error creating output directory
* NopolExecStrat - spoon relies on class path to work but this does not work very well in Jenkins context
 * package instrumenting does not exist
* AstorJKali
 * log-4j class not found - banned on jenkins (work-around with maven option `-Denforcer.skip=true` does not work)
* AstorJMut
 * Same error as AstorJKali

Conclusion: development and mantainance cost for the plugin is very expensive due to 
* Dependency clash with Jenkins required dependency
* due to dependency clashes and transitive imports failures , tests can't be automated
* Need to keep 2 version of the same thing - a repairnator for Jenkins and a repairnator for the usual context
* Repairnator strategy is to integrate repair tools with each other so the plugin might not be very scalable according to the number of new tools - due to some tools might be impossible to be running in Jenkins context like `NopolExecStrat` 

Better solution is to use Docker , which we already implemented and it is much cheaper and easier to maintain (we only need to test the usual repairnator) than importing Repairnator as a maven dependency . So I propose we only need to at most develop a Jenkins plugin interface to simplify the usage for the users but it in fact execute docker or alternatively somehow show the user how to use our docker image instead. That's all for my findings.

**Comment by monperrus** [comment]
great wrap-up. to be discussed. the logging issues should be easy to fix.

the problem with docker is that very few Jenkins have nodes with docker installed.

another option is to run the repair attempts in a separate JVM (all Jenkins nodes hava Java): no
dependency problems, no docker requirements

**Comment by henry-lp** [comment]
Thanks @monperrus , great tips. In that case , running a separate JVM through subprocess would be better than docker as you suggested. I'll try and see if I can resolve these issue or find som kind of work around otherwise I would go with this suggestion .

**Comment by henry-lp** [comment]
Another problem is that Repairnator use Maven during run so we need to specify that every node must have maven installed in order to run repairnator - this is not related to any of the errors above.

**Comment by monperrus** [comment]
Not necessarily. If we have Maven in the pipeline-jar-with-dependencies, we don't need to have it
installed on the node.

**Comment by henry-lp** [comment]
Last time I ran the plugins (repairnator as dependency , not the subprocess one) on Long's Jenkins server . It complained about not being able to run maven and crashed during testInformationGathering step. Maybe this is not the case if we run as subproc as you said . I'll test this out later on his server again and come back here with a report.

**Comment by monperrus** [comment]
there may be some changes to be done for that, such as adding a dependency to maven jars in the
pipeline pom.xml

**Comment by henry-lp** [comment]
I'll investigate more on that, it should be solvable.

**Comment by henry-lp** [comment]
Currently I am developing 2 version of the plugin . The first relies on pure java as maven import and the other will exec the downloaded jar file as a sub process (separate JVM) .

* Pure Java Problems
  * AstorJKali and AstorJMut does not work due to the Logger class of log4j used for AstorOutput clashed with Jenkins plugin parent dependency log4j-over-slf4j , when exec Logger.getRepository it assume the the Logger is of class Logger from log4j-over-slf4j - the biggest problem is that we aren't allowed to exclude log4j-over-slf4j from parent pom, which always load its dependencies first.
  * AssertFixer - infinite loop , AssertFixer try scheduling its threads via executor service but the thread got infinitely delayed due to ? access right or jenkins does not have enough cpu resource to run it ? . Previously Assertfixer had a classdefnot found problem but it was fixed by including a dependency.
  * NopolExecStrat - [detailed Investigation in prog]

* As Subprocess Problem
  * Only problem with AssertFixer (same as pure java , infinite run) the other tools run fine , NopolExecStrat will be tested tommorow and update this message.

**Comment by monperrus** [comment]
Great. I think subprocess is less brittle and more maintainable in the long term. It would also
allow non-java repair tools to be used such as Sequencer.

**Comment by monperrus** [comment]
Per our discussion today, we don't support AssertFixer in the Jenkins plugin (because of some internal incompatibility, probably because of the usage of ExecutorService in AssertFixer).

**Comment by henry-lp** [comment]
@monperrus it says that I need you to +1 the request , here is the [link](https://bugs.eclipse.org/bugs/show_bug.cgi?id=559662) . 

[How to request a JIPP](https://wiki.eclipse.org/CBI#Requesting_a_JIPP_instance) under ``Requesting a JIPP instance`

**Comment by monperrus** [comment]
> https://ci.eclipse.org/repairnator/

Great! Works for me also! 

Next step is to have a working version of the Repairnator plugin on the Repairnator CI!

**Comment by henry-lp** [comment]
@monperrus I can do that . Can you elevate my access in https://ci.eclipse.org/repairnator/ , since I am currently not allowed to installing `.hpi` there .

**Comment by monperrus** [comment]
I can see you https://ci.eclipse.org/repairnator/user/tailp@kth.se/ but I cannot find the way to
give you this right. Could you ask support? Thanks!

**Comment by henry-lp** [comment]
@monperrus hi Martin , I just checked the issue thread [link](https://bugs.eclipse.org/bugs/show_bug.cgi?id=559662) . It seems to be some troublesome requirements, but no matter I currently am already testing this on Long's server at `jenkins.kth-assert.net` and in progressing fixing the maven dependency issue of our repairnator.

**Comment by henry-lp** [comment]
Maven issue is fixed using Jenkins standard build-in maven installer.

**Comment by monperrus** [comment]
cool!

**Comment by henry-lp** [comment]
@monperrus 

After PR (#1019) is merged, I'll make a video as the last step and we are more or less finished with this task . 

What could have done more:
* 4 more happy test cases for (AstorJKali, AstorJMut,NPEFixSafe,NopolExec) - currently there is only one auto test for NPEFix

**Comment by monperrus** [comment]
closed per #1019

**Comment by monperrus** [comment]
Follow-up: #1042
